src/Internal/Payment/Storage.php

```diff
@@ -0,0 +1,175 @@
+<?php
+/**
+ * Class Storage
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment;
+
+use WC_Order;
+use WCPay\Internal\Payment\State\InitialState;
+use WCPay\Internal\Payment\State\State;
+use WCPay\Internal\Service\PaymentMethodService;
+
+/**
+ * Storage for payments.
+ */
+class Storage {
+	/**
+	 * Key for storing data in order meta.
+	 *
+	 * @var string
+	 */
+	const ORDER_META_KEY = '_wcpay_payment_data';
+
+	/**
+	 * State factory.
+	 *
+	 * @var StateFactory
+	 */
+	private $state_factory;
+
+	/**
+	 * Payment methods service.
+	 *
+	 * @var PaymentMethodService
+	 */
+	private $payment_method_service;
+
+	/**
+	 * Controls whether storage is actually enabled.
+	 *
+	 * @var bool
+	 */
+	private $enabled = false;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param StateFactory         $state_factory          A factory for payment states.
+	 * @param PaymentMethodService $payment_method_service PM service.
+	 */
+	public function __construct(
+		StateFactory $state_factory,
+		PaymentMethodService $payment_method_service
+	) {
+		$this->state_factory          = $state_factory;
+		$this->payment_method_service = $payment_method_service;
+	}
+
+	/**
+	 * Enables payment storage.
+	 *
+	 * IMPORTANT: By default, this class does not touch the database.
+	 * Once enabled, it will, but this should only be used for tests
+	 * before the final storage method is determined.
+	 */
+	public function enable() {
+		$this->enabled = true;
+	}
+
+	/**
+	 * Loads or creates the payment object for an order,
+	 * returning the one true payment object.
+	 *
+	 * @param WC_Order $order Order that requires/has payment.
+	 * @return State          Current state of the order payment.
+	 */
+	public function get_order_payment( WC_Order $order ) {
+		$payment     = new Payment( $order );
+		$state_class = InitialState::class;
+
+		if ( $this->enabled ) {
+			$data = $order->get_meta( self::ORDER_META_KEY );
+			if ( is_array( $data ) && ! empty( $data ) ) {
+				$state_class = $data['state'];
+				$this->import_payment_data( $payment, $data );
+			}
+		}
+
+		$state = $this->state_factory->create_state( $state_class );
+		$state->set_context( $payment );
```

Comment:
This is how I imagine states being created, at least when we're always starting with `InitialState` or loading payments from storage.